go-controller/pkg/ovn/pods.go

```diff
@@ -44,8 +45,22 @@ func (oc *Controller) syncPods(pods []interface{}) error {
 			}
 			logicalPort := util.GetLogicalPortName(pod.Namespace, pod.Name)
 			expectedLogicalPorts[logicalPort] = true
+			// it is possible to try to add a pod here that has no node. For example if a pod was deleted with
+			// a finalizer, and then the node was removed. In this case the pod will still exist in a running state.
+			if _, err := oc.watchFactory.GetNode(pod.Spec.NodeName); kerrors.IsNotFound(err) &&
+				oc.lsManager.GetSwitchSubnets(pod.Spec.NodeName) == nil {
+				if util.PodTerminatingWithFinalizer(pod) {
+					klog.Infof("Ignoring IP allocation for pod terminating with finalizer: %s/%s, on deleted "+
+						"node: %s", pod.Namespace, pod.Name, pod.Spec.NodeName)
+					continue
+				} else {
+					// unknown condition how we are getting a running pod without a node here
+					klog.Errorf("Pod IP allocation found for a non-existent node in API with unknown "+
+						"condition. Pod: %s/%s, node: %s", pod.Namespace, pod.Name, pod.Spec.NodeName)
+				}
+			}
 			if err = oc.waitForNodeLogicalSwitchInCache(pod.Spec.NodeName); err != nil {
-				return fmt.Errorf("failed to wait for node %s to be added to cache. IP allocation may fail!",
+				return fmt.Errorf("failed to wait for node %s to be added to cache. IP allocation may fail",
```

Comment:
So what we are basically doing is that if the node doesn't exist and the pod is finalizing, we assume that the node was deleted. Otherwise, we assume that if the node doesn't exist it is an error.

Would it make more sense to handle it within this if? Could we be in the same situation with the pod in a different state? Like `PodCompleted`?

I am also confused about us doing `waitForNodeLogicalSwitchInCache` here, but I might be missing something. We do `WatchNodes` before `WatchPods`, and the sync functions are synchronous so I am not sure what are we waiting for.

Using `oc.watchFactory.GetNode(pod.Spec.NodeName)` might be racey if the node was deleted concurrently.

Anyway, I would probably replace it all with a new error for `AllocateIPs`, such as `ErrNodeUnknown`, and then ignore the error if the pod is terminating or terminated. I find it more expressive. An if we think we need to wait, just to be prepared if this is ever asynchronous or if I did miss something, replace that with a `retry.OnError` of AllocateIPs.

This might be a nit, up to you.

Reply:
I'll look into the pod completed case, but without a finalizer I think it would be removed when the node is removed. Let me get back to you on that.

For waitForNodeLogicalSwitchInCache, we cannot assume because WatchNodes started before WatchPods that all nodes have been added. For example, this could happen:
1. WatchNodes() executes
2. a new node with a pod is added
3. WatchPods() executes - here we can wait for the node and ip allocator be added by the node watcher thread

Yeah it could be racey, but if hte node was deleted and then added back concurrently, the node code path should add the pod allocations back. I'll check on this code path too.

I wasn't really changing the error here, just fixing a gofmt thing with punctuation at the end of the error.

Reply:
Actually for pods that are completed we already ignore them in the sync on line 41.

Reply:
Forget about what I said related `waitForNodeLogicalSwitchInCache`, for some reason I thought no worker threads would be running while we were syncing.

The code though still feels to me more complicated that it needs to be. The perspective that I have for this code is: Let's just try allocating the IPs and retry on certain conditions, continue on others and fail for the rest.

But this might be just my nit.

Reply:
> Actually for pods that are completed we already ignore them in the sync on line 41.

Yeah, but it is not very intuitive why we are ignoring completed pods, but not pods terminating with a finalizer. They both technically exist. just confusing though, I don't see anything necessarily wrong.